elfo-network/src/connman.rs

```diff
@@ -0,0 +1,366 @@
+            readvise_at: Instant::now(),
+        }
+    }
+
+    pub(crate) fn drain_task_queue(
+        &mut self,
+    ) -> (
+        Option<Duration>,
+        impl ExactSizeIterator<Item = Task> + use<'_>,
+    ) {
+        let now = Instant::now();
+
+        macro_rules! transition {
+            ($conn_id:expr, $From:ident => $To:ident) => {{
+                self.conns.get_mut($conn_id).and_then(|c| {
+                    if matches!(c.state(), State::$From) {
+                        Some((c, State::$To))
+                    } else {
+                        None
+                    }
+                })
+            }};
+        }
+
+        let Some(mut cur_idx) = self.task_queue.len().checked_sub(1) else {
+            // It is empty anyway.
+            return (None, self.tq_iter_buffer.drain(..));
+        };
+
+        let mut least_time: Option<Instant> = None;
+        loop {
+            let task = &self.task_queue[cur_idx];
+            let correct = match &task.data {
+                TaskData::ConnectNew(id) => transition!(*id, New => Establishing),
+                TaskData::ReconnectFailed(id) => transition!(*id, Failed => Establishing),
+            };
+
+            if let Some((conn, next_state)) = correct {
+                let diff = task.at.duration_since(now);
+                // tokio timers are 1ms precise anyway.
+                if diff.as_millis() <= 1 {
+                    let task = self.task_queue.swap_remove(cur_idx);
+                    self.tq_iter_buffer.push(task);
+                    conn.state = next_state;
+                } else if let Some(least) = &mut least_time {
+                    *least = (*least).min(task.at);
+                } else {
+                    least_time = Some(task.at);
+                }
+            } else {
+                self.task_queue.swap_remove(cur_idx);
+            }
+
+            if let Some(next) = cur_idx.checked_sub(1) {
+                cur_idx = next;
+            } else {
+                break;
+            }
+        }
+
+        // Do not advise as long as previous advise is in progress.
+        let next_check = least_time
+            .and_then(|next_check| {
+                if now >= self.readvise_at {
+                    self.readvise_at = next_check;
+                    Some(next_check)
+                } else {
+                    None
+                }
+            })
+            .map(|at| at.duration_since(now));
+        (next_check, self.tq_iter_buffer.drain(..))
+    }
+
+    pub(crate) fn insert_new(
+        &mut self,
+        role: ConnectionRole,
+        transport: ConnectTransport,
+    ) -> ManagedConn<'_> {
+        let id = self.conns.insert(Conn {
+            role,
+            transport,
+            state: State::New,
+        });
+        self.task_queue
+            .push(Task::new(Instant::now(), TaskData::ConnectNew(id)));
+        // SAFETY: just inserted.
+        unsafe { self.get_mut(id).unwrap_unchecked() }
```

Comment:
Let's use `expect()` because we don't have enough tests to detect regression, and safety guarantees are across several methods.